Src/library/net/network/socket/network_socket.e

```diff
@@ -311,31 +381,49 @@ feature {NONE} -- Implementation
 
 feature {NONE} -- Constants
 
-	default_timeout: INTEGER = 20
+	default_timeout: INTEGER
 			-- Default timeout duration in seconds
 
-feature {NONE} -- Externals
-
-	c_set_sock_recv_timeout (a_fd, level: INTEGER; a_timeout_seconds: INTEGER)
-			-- C routine to set socket option `SO_RCVTIMEO' with `a_timeout_seconds' seconds.
-		external
-			"C"
+	one_second_in_nanoseconds: NATURAL_64 = 1_000_000_000
+
+	max_timeout_ns_value: NATURAL_64 = 2_147_483_647_999_999_999
+			-- Maximum value for a timeout in nanoseconds.
+			-- See `is_valid_timeout_ns` for explanation.
+
+feature -- Validation			
+
+	is_valid_timeout_ns (ns: NATURAL_64): BOOLEAN
+			--| note: internally (C API), the ns timeout is splitted into:
+			--|  - a long value `tv_sec`: number of seconds
+			--|  - a long value `tv_usec`: number of microseconds (with struct timeval)
+			--|  or  long value `tv_nsec`: number of nanoseconds  (with struct timespec)
+			--| tv_usec or tv_nsec should be >= 0 and < equivalent of 1 second.
+			--| tv_usec < 1_000_000
+			--| tv_nsec < 1_000_000_000
+			--| so there is no issue with NATURAL_64 for usec or nsec
+			--| but, then, there is restriction for the seconds value which is coded as long int
+			--|  and long values (32 bits) are between -2_147_483_647 and +2_147_483_647 (i.e 2e31 - 1 = {INTEGER_32}.max_value).
+			--|  then, the `seconds` part of the timeout <= +2_147_483_647, and then expressed in nanoseconds, and taking into account
+			--|  the usec or nsec part, the timeout should not be greater than:
+			--|     +2_147_483_647 * 1_000_000_000 + 999_999_999 =  2_147_483_647_999_999_999
+			--|      and as NATURAL_64's max value is bigger     = 18_446_744_073_709_551_615
+			--|      features dealing with timeout in nanoseconds have related preconditions.
+		do
+			Result := 0 < ns and then ns <= max_timeout_ns_value
 		end
 
-	c_get_sock_recv_timeout (a_fd, level: INTEGER): INTEGER
-			-- C routine to get socket option SO_RCVTIMEO of timeout value in seconds.
-		external
-			"C"
-		end
+feature {NONE} -- Externals
 
-	c_set_sock_send_timeout (a_fd, level: INTEGER; a_timeout_seconds: INTEGER)
-			-- C routine to set socket option `SO_SNDTIMEO' with `a_timeout_seconds' seconds.
+	c_set_sock_opt_timeo (a_fd, level: INTEGER; a_timeo_opt: INTEGER; a_timeout_nanoseconds: NATURAL_64)
+		-- C routine to set socket option `$a_timeo_opt' with `a_timeout_nanoseconds`.
+		require
+			is_valid_timeout_ns: is_valid_timeout_ns (a_timeout_nanoseconds)
```

Comment:
I guess, there are also restrictions on `level` and `a_timeo_opt`.

Reply:
I guess so as well.
This would be a part of another patch.
It is better to have small patch, rather than big one.